File: backend/src/deaths_door/routes/lights.py
"""API routes for DMX lighting control."""

from typing import Any

from fastapi import APIRouter, HTTPException, Path
from pydantic import BaseModel, Field

from ..lighting import LightingManager, LightingScene
from ..sound_fx import SoundFX, SoundName

router = APIRouter(prefix="/lights", tags=["Lighting"])


@router.get("/status")
async def get_status() -> dict[str, Any]:
    """Get DMX connection status and detected devices."""
    manager = LightingManager()
    
    # Get serial port info if connected
    port_info = None
    if manager.controller and manager.controller.serial_port:
        port_info = {
            "port": manager.controller.serial_port.port,
            "is_open": manager.controller.serial_port.is_open,
        }
    
    return {
        "connected": manager.connected,
        "serial_port": port_info,
        "fixtures": {
            "light1": manager.light1 is not None,
            "light2": manager.light2 is not None,
            "fog": manager.fog is not None,
        },
        "calibrated_positions": len(manager.positions),
    }


# Response Models
class OperationResponse(BaseModel):
    """Response for successful operations."""

    status: str = Field(..., description="Operation status", examples=["success"])
    message: str = Field(
        ..., description="Operation message", examples=["Lights updated"]
    )


class SceneListResponse(BaseModel):
    """Response with list of available scenes."""

    scenes: list[str] = Field(..., description="List of available scene names")


class PositionResponse(BaseModel):
    """Response with player position information."""

    player_num: int = Field(..., description="Player number")
    pan: int = Field(..., description="Pan value (0-255)")
    tilt: int = Field(..., description="Tilt value (0-255)")


class PositionsListResponse(BaseModel):
    """Response with all saved positions."""

    positions: dict[int, dict[str, int]] = Field(
        ..., description="Dictionary of player positions"
    )


# Request Models
class ChannelSetRequest(BaseModel):
    """Request to set a channel value."""

    value: int = Field(..., description="Channel value (0-255)", ge=0, le=255)


class PositionSetRequest(BaseModel):
    """Request to set pan/tilt position."""

    pan: int = Field(..., description="Pan value (0-255)", ge=0, le=255)
    tilt: int = Field(..., description="Tilt value (0-255)", ge=0, le=255)
    fine: bool = Field(True, description="Use fine adjustment")


class CalibrationSaveRequest(BaseModel):
    """Request to save calibrated position."""

    pan: int = Field(..., description="Pan value (0-255)", ge=0, le=255)
    tilt: int = Field(..., description="Tilt value (0-255)", ge=0, le=255)


class SpotlightRequest(BaseModel):
    """Request to spotlight a player."""

    brightness: int = Field(
        255, description="Brightness level (0-255)", ge=0, le=255
    )
    fixture_id: int = Field(1, description="Fixture ID to use (1 or 2)", ge=1, le=2)


# Scene Control Endpoints
@router.get("/scenes/list")
async def list_scenes() -> SceneListResponse:
    """List all available lighting scenes."""
    manager = LightingManager()
    return SceneListResponse(scenes=manager.list_scenes())


@router.post(
    "/scene/{name}",
    responses={
        404: {"description": "Scene not found"},
        500: {"description": "Failed to trigger scene"},
    },
)
async def trigger_scene(
    name: str = Path(..., description="Name of the scene to trigger", examples=["death"])
) -> OperationResponse:
    """
    Trigger a lighting scene.

    This endpoint triggers only the lighting portion of a scene.
    For integrated light+sound scenes, use the /scenes/integrated/{name} endpoint.
    """
    scene = LightingScene.from_str(name)
    if scene is None:
        raise HTTPException(status_code=404, detail=f"Scene '{name}' not found")

    try:
        manager = LightingManager()
        manager.trigger_scene(name)
        return OperationResponse(
            status="success", message=f"Scene '{name}' triggered successfully"
        )
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to trigger scene: {str(e)}"
        ) from e


@router.post(
    "/scene/integrated/{name}",
    responses={
        404: {"description": "Scene not found"},
        500: {"description": "Failed to trigger scene"},
    },
)
async def trigger_integrated_scene(
    name: str = Path(
        ...,
        description="Name of the integrated scene to trigger",
        examples=["death"],
    )
) -> OperationResponse:
    """
    Trigger an integrated light+sound scene.

    This endpoint triggers both lighting and sound effects simultaneously
    for a coordinated audio-visual experience.
    """
    # Check if scene exists for lighting
    lighting_scene = LightingScene.from_str(name)
    if lighting_scene is None:
        raise HTTPException(status_code=404, detail=f"Scene '{name}' not found")

    # Check if sound exists
    sound_name = SoundName.from_str(name)

    try:
        # Trigger lighting
        lighting_manager = LightingManager()
        lighting_manager.trigger_scene(name)

        # Trigger sound if available
        if sound_name is not None:
            sound_fx = SoundFX()
            sound_fx.play(sound_name)

        return OperationResponse(
            status="success",
            message=f"Integrated scene '{name}' triggered successfully",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to trigger scene: {str(e)}"
        ) from e


# Granular Control Endpoints
@router.post("/fixture/{fixture_id}/channel/{channel}")
async def set_channel(
    fixture_id: int = Path(..., description="Fixture ID (1=Light1, 2=Light2, 3=Fog)", ge=1, le=3),
    channel: int = Path(..., description="Channel number (1-11)", ge=1, le=11),
    request: ChannelSetRequest = ...,
) -> OperationResponse:
    """
    Set a specific DMX channel value on a fixture.

    This provides granular control for manual adjustments and testing.

    Channel mapping for XPCLEOYZ Moving Head (11-channel mode):
    - 1: Pan Running
    - 2: Pan Fine
    - 3: Tilt Running
    - 4: Tilt Fine
    - 5: Color
    - 6: Gobo
    - 7: Fixed Gobo/Pattern
    - 8: Strobe
    - 9: Dimming
    - 10: Auto Mode
    - 11: Pan/Tilt Running/Reset
    """
    try:
        manager = LightingManager()
        manager.set_channel(fixture_id, channel, request.value)
        return OperationResponse(
            status="success",
            message=f"Fixture {fixture_id} channel {channel} set to {request.value}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to set channel: {str(e)}"
        ) from e


@router.post("/fixture/{fixture_id}/position")
async def set_position(
    fixture_id: int = Path(..., description="Fixture ID (1 or 2)", ge=1, le=2),
    request: PositionSetRequest = ...,
) -> OperationResponse:
    """
    Set pan/tilt position of a moving head fixture.

    This is a convenience endpoint for positioning lights without
    dealing with individual channels.
    """
    try:
        manager = LightingManager()
        manager.set_position(fixture_id, request.pan, request.tilt, request.fine)
        return OperationResponse(
            status="success",
            message=f"Fixture {fixture_id} positioned to pan={request.pan}, tilt={request.tilt}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to set position: {str(e)}"
        ) from e


@router.post("/blackout")
async def blackout() -> OperationResponse:
    """
    Turn off all lights immediately.

    This is an emergency blackout function that sets all dimmer channels to 0.
    """
    try:
        manager = LightingManager()
        manager.blackout()
        return OperationResponse(status="success", message="All lights blacked out")
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to blackout: {str(e)}"
        ) from e


# Calibration Endpoints
@router.post("/calibrate/player/{player_num}/save")
async def save_player_position(
    player_num: int = Path(..., description="Player number (chair position)", ge=1),
    request: CalibrationSaveRequest = ...,
) -> OperationResponse:
    """
    Save a calibrated position for a specific player.

    Use this after manually adjusting lights to point at a player's chair.
    The position can then be recalled quickly using the spotlight endpoint.

    Workflow:
    1. Use /fixture/{id}/position to manually adjust lights
    2. Once positioned correctly, call this endpoint to save the position
    3. Use /lights/spotlight/player/{num} to recall the saved position
    """
    try:
        manager = LightingManager()
        manager.save_player_position(player_num, request.pan, request.tilt)
        return OperationResponse(
            status="success",
            message=f"Position saved for player {player_num}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to save position: {str(e)}"
        ) from e


@router.get("/calibrate/positions")
async def get_all_positions() -> PositionsListResponse:
    """
    Get all saved player positions.

    Returns a dictionary mapping player numbers to their calibrated pan/tilt positions.
    """
    try:
        manager = LightingManager()
        positions = manager.get_all_positions()
        return PositionsListResponse(positions=positions)
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to get positions: {str(e)}"
        ) from e


@router.post("/spotlight/player/{player_num}")
async def spotlight_player(
    player_num: int = Path(..., description="Player number to spotlight", ge=1),
    request: SpotlightRequest = SpotlightRequest(),
) -> OperationResponse:
    """
    Spotlight a specific player using their saved calibrated position.

    This will move the specified fixture to the saved position for the player
    and activate spotlight mode (white light, full brightness by default).

    The player position must be calibrated first using the calibration endpoints.
    """
    try:
        manager = LightingManager()

        # Check if position exists
        if player_num not in manager.positions:
            raise HTTPException(
                status_code=404,
                detail=f"No calibrated position found for player {player_num}. "
                "Please calibrate the position first.",
            )

        manager.spotlight_player(
            player_num, request.brightness, request.fixture_id
        )
        return OperationResponse(
            status="success",
            message=f"Spotlighting player {player_num} with fixture {request.fixture_id}",
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to spotlight player: {str(e)}"
        ) from e

File: backend/src/routes/lights.rs
//! API routes for DMX lighting control.

use std::collections::BTreeMap;

use axum::extract::{Path, Query, State};
use axum::Json;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use utoipa::ToSchema;
use utoipa_axum::router::OpenApiRouter;
use utoipa_axum::routes;

use crate::app::{AppError, AppJson, AppResult, AppState};
use crate::effects::{paired_sound, ActiveEffect};
use crate::lighting::LightingScene;

pub fn router() -> OpenApiRouter<AppState> {
    OpenApiRouter::new()
        .routes(routes!(get_status))
        .routes(routes!(list_scenes))
        .routes(routes!(trigger_scene))
        .routes(routes!(trigger_integrated_scene))
        .routes(routes!(set_channel))
        .routes(routes!(set_position))
        .routes(routes!(blackout))
        .routes(routes!(save_player_position))
        .routes(routes!(get_all_positions))
        .routes(routes!(spotlight_player))
}

#[derive(Serialize, ToSchema)]
pub struct OperationResponse {
    pub status: String,
    pub message: String,
}

fn ok(message: impl Into<String>) -> Json<OperationResponse> {
    Json(OperationResponse {
        status: "success".to_string(),
        message: message.into(),
    })
}

#[derive(Serialize, ToSchema)]
pub struct SceneListResponse {
    pub scenes: Vec<String>,
}

#[utoipa::path(
    get, path = "/lights/status", tag = "Lighting",
    responses((status = 200, description = "DMX connection status and detected devices"))
)]
async fn get_status(State(state): State<AppState>) -> Json<Value> {
    let status = state.lighting.status();
    let port_info = status
        .serial_port
        .map(|port| json!({ "port": port, "is_open": true }))
        .unwrap_or(Value::Null);
    Json(json!({
        "connected": status.connected,
        "serial_port": port_info,
        "fixtures": {
            "light1": status.has_light1,
            "light2": status.has_light2,
            "fog": status.has_fog,
        },
        "calibrated_positions": status.calibrated_positions,
    }))
}

#[utoipa::path(
    get, path = "/lights/scenes/list", tag = "Lighting",
    responses((status = 200, description = "Available scene names", body = SceneListResponse))
)]
async fn list_scenes(State(state): State<AppState>) -> Json<SceneListResponse> {
    Json(SceneListResponse {
        scenes: state.lighting.list_scenes(),
    })
}

#[derive(Deserialize, ToSchema)]
pub struct SceneQuery {
    /// Skip the scene's paired sound effect (lights and overlay only).
    #[serde(default)]
    pub silent: bool,
}

#[derive(Serialize, ToSchema)]
pub struct SceneTriggerResponse {
    pub status: String,
    /// The effect now playing (its id, scene, and duration). The same object
    /// appears in `/game/state` and SSE frames while it plays.
    pub effect: ActiveEffect,
    /// The paired sound that was played, if any.
    pub sound: Option<String>,
}

/// Trigger a scene: timed lighting sequence + paired sound + overlay visual.
///
/// The effect's length follows its sound's audio duration (e.g. the death sting
/// is ~1.5s; the goodnight music box is ~13s). Pass `?silent=true` to skip the
/// sound. Scene/sound pairs: death→death, drama→drama, goodnight→music_box,
/// morning→rooster, reveal→drumroll.
#[utoipa::path(
    post, path = "/lights/scene/{name}", tag = "Lighting",
    params(
        ("name" = String, Path, description = "Scene name, e.g. death/drama/blackout"),
        ("silent" = Option<bool>, Query, description = "Skip the paired sound effect")
    ),
    responses(
        (status = 200, description = "Scene triggered", body = SceneTriggerResponse),
        (status = 404, description = "Scene not found")
    )
)]
async fn trigger_scene(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Query(query): Query<SceneQuery>,
) -> AppResult<Json<SceneTriggerResponse>> {
    let scene = LightingScene::from_str(&name)
        .ok_or_else(|| AppError::not_found(format!("Scene '{name}' not found")))?;
    let sound = if query.silent {
        None
    } else {
        paired_sound(scene)
    };
    let effect = state.effects.trigger(scene, query.silent).await;
    Ok(Json(SceneTriggerResponse {
        status: "success".to_string(),
        effect,
        sound: sound.map(|s| s.value().to_string()),
    }))
}

/// Deprecated alias for `POST /lights/scene/{name}`, which now plays the paired
/// sound by default.
#[utoipa::path(
    post, path = "/lights/scene/integrated/{name}", tag = "Lighting",
    params(("name" = String, Path, description = "Scene name")),
    responses(
        (status = 200, description = "Scene triggered", body = SceneTriggerResponse),
        (status = 404, description = "Scene not found")
    )
)]
async fn trigger_integrated_scene(
    state: State<AppState>,
    name: Path<String>,
) -> AppResult<Json<SceneTriggerResponse>> {
    trigger_scene(state, name, Query(SceneQuery { silent: false })).await
}

#[derive(Deserialize, ToSchema)]
pub struct ChannelSetRequest {
    pub value: i64,
}

#[utoipa::path(
    post, path = "/lights/fixture/{fixture_id}/channel/{channel}", tag = "Lighting",
    params(
        ("fixture_id" = i64, Path, description = "1=Light1, 2=Light2, 3=Fog"),
        ("channel" = i64, Path, description = "DMX channel 1-11")
    ),
    request_body = ChannelSetRequest,
    responses((status = 200, description = "Channel set", body = OperationResponse))
)]
async fn set_channel(
    State(state): State<AppState>,
    Path((fixture_id, channel)): Path<(i64, i64)>,
    AppJson(req): AppJson<ChannelSetRequest>,
) -> AppResult<Json<OperationResponse>> {
    state.lighting.set_channel(fixture_id, channel, req.value);
    Ok(ok(format!(
        "Fixture {fixture_id} channel {channel} set to {}",
        req.value
    )))
}

#[derive(Deserialize, ToSchema)]
pub struct PositionSetRequest {
    pub pan: i64,
    pub tilt: i64,
    #[serde(default = "default_true")]
    pub fine: bool,
}

fn default_true() -> bool {
    true
}

#[utoipa::path(
    post, path = "/lights/fixture/{fixture_id}/position", tag = "Lighting",
    params(("fixture_id" = i64, Path, description = "1 or 2")),
    request_body = PositionSetRequest,
    responses((status = 200, description = "Fixture positioned", body = OperationResponse))
)]
async fn set_position(
    State(state): State<AppState>,
    Path(fixture_id): Path<i64>,
    AppJson(req): AppJson<PositionSetRequest>,
) -> AppResult<Json<OperationResponse>> {
    state
        .lighting
        .set_position(fixture_id, req.pan, req.tilt, req.fine);
    Ok(ok(format!(
        "Fixture {fixture_id} positioned to pan={}, tilt={}",
        req.pan, req.tilt
    )))
}

#[utoipa::path(
    post, path = "/lights/blackout", tag = "Lighting",
    responses((status = 200, description = "All lights blacked out", body = OperationResponse))
)]
async fn blackout(State(state): State<AppState>) -> AppResult<Json<OperationResponse>> {
    state.lighting.blackout();
    Ok(ok("All lights blacked out"))
}

#[derive(Deserialize, ToSchema)]
pub struct CalibrationSaveRequest {
    pub pan: i64,
    pub tilt: i64,
}

#[utoipa::path(
    post, path = "/lights/calibrate/player/{player_num}/save", tag = "Lighting",
    params(("player_num" = i64, Path, description = "Player/chair number")),
    request_body = CalibrationSaveRequest,
    responses((status = 200, description = "Position saved", body = OperationResponse))
)]
async fn save_player_position(
    State(state): State<AppState>,
    Path(player_num): Path<i64>,
    AppJson(req): AppJson<CalibrationSaveRequest>,
) -> AppResult<Json<OperationResponse>> {
    state
        .lighting
        .save_player_position(player_num, req.pan, req.tilt);
    Ok(ok(format!("Position saved for player {player_num}")))
}

#[derive(Serialize, ToSchema)]
pub struct PositionsListResponse {
    pub positions: BTreeMap<i64, BTreeMap<String, i64>>,
}

#[utoipa::path(
    get, path = "/lights/calibrate/positions", tag = "Lighting",
    responses((status = 200, description = "All calibrated positions", body = PositionsListResponse))
)]
async fn get_all_positions(State(state): State<AppState>) -> Json<PositionsListResponse> {
    let positions = state
        .lighting
        .get_all_positions()
        .into_iter()
        .map(|(num, pos)| {
            let mut m = BTreeMap::new();
            m.insert("player_num".to_string(), pos.player_num);
            m.insert("pan".to_string(), pos.pan);
            m.insert("tilt".to_string(), pos.tilt);
            (num, m)
        })
        .collect();
    Json(PositionsListResponse { positions })
}

#[derive(Deserialize, ToSchema)]
pub struct SpotlightRequest {
    #[serde(default = "default_brightness")]
    pub brightness: i64,
    #[serde(default = "default_fixture")]
    pub fixture_id: i64,
}

fn default_brightness() -> i64 {
    255
}
fn default_fixture() -> i64 {
    1
}

#[utoipa::path(
    post, path = "/lights/spotlight/player/{player_num}", tag = "Lighting",
    params(("player_num" = i64, Path, description = "Player number")),
    request_body = SpotlightRequest,
    responses(
        (status = 200, description = "Player spotlighted", body = OperationResponse),
        (status = 404, description = "No calibrated position for that player")
    )
)]
async fn spotlight_player(
    State(state): State<AppState>,
    Path(player_num): Path<i64>,
    AppJson(req): AppJson<SpotlightRequest>,
) -> AppResult<Json<OperationResponse>> {
    if !state.lighting.has_position(player_num) {
        return Err(AppError::not_found(format!(
            "No calibrated position found for player {player_num}. Please calibrate the position first."
        )));
    }
    state
        .lighting
        .spotlight_player(player_num, req.brightness, req.fixture_id);
    Ok(ok(format!(
        "Spotlighting player {player_num} with fixture {}",
        req.fixture_id
    )))
}
